python/tank/templatekey.py

```diff
@@ -256,6 +273,145 @@ def _as_string(self, value):
+        """
+        if isinstance(format_spec, basestring) is False:
+            raise TankError("format_spec for TimestampKey %s is not of type string: %s" %
+                            (name, format_spec.__class__.__name__))
+        self.format_spec = format_spec
+
+        if isinstance(utc_default, bool) is False:
+            raise TankError("utc_default for TimestampKey %s is not of type boolean: %s" %
+                            (name, utc_default.__class__.__name__))
+        self.utc_default = utc_default
+
+        super(TimestampKey, self).__init__(
+            name,
+            default=self.__get_current_time
+        )
+
+    def __get_current_time(self):
+        """
+        Returns the current time as a datetime.datetime instance.
+
+        Do not streamline the code so the __init__ method simply passesd the dt.datetime.now method,
+        we can't mock datetime.now since it's builtin and will make unit tests more complicated to
+        write.
+
+        :returns: If utc_default, a datetime object representing time in the UTC timezone. Otherwise,
+                  a datetime object representing time in the local timezone. In any case, the tzinfo
+                  member is None.
+        """
+        if self.utc_default:
+            return dt.datetime.utcnow()
+        else:
+            return dt.datetime.now()
+
+    def validate(self, value):
+        """
+        Test if a value is valid for this key.
+
+        :param value: Value to test.
+
+        :returns: Bool
+        """
+        if isinstance(value, basestring) or isinstance(value, unicode):
+            return self._validate_str(value)
+        else:
+            return self._validate_value(value)
+
+    def _validate_str(self, str_value):
+        """
+        Test if a string is valid for this key.
+
+        :param str_value: String to test.
+
+        :returns: Bool
+        """
+        try:
+            self._as_value(str_value)
+            return True
+        except ValueError, e:
+            self._last_error = "Invalid string: %s" % e.message
+            return False
+        except TypeError, e:
+            self._last_error = "Invalid type: %s" % e.message
+            return False
+
+    def _validate_value(self, value):
+        """
+        Test if a string is valid for this key.
+
+        :param str_value: String to test.
+
+        :returns: Bool
+        """
+        try:
+            self._as_string(value)
+        except TypeError, e:
+            self._last_error = "Invalid type: %s" % e.message
+            return False
+        except ValueError:
+            self._last_error = "Invalid value: %s" % e.message
+            return False
+        return True
+
+    def _as_string(self, value):
+        """
+        Converts a given value as string.
+
+        :param value: Can be any of:
+                          - int or float representing seconds since epoch in local time.
+                          - datetime.datetime
```

Comment:
I think we are maybe offering too many options here. With other keys, like the int template key, we offer the user to cast from one data type (int) to a string and back. I would perfer if we preserved that rationale here and essentially allowed a user to go from a string to a datetime and back. This is the symmetry of the class as far as I see it. You should be able to put something in and then get it back out again. So I think it's better if this method only accepts datetime objects(because that's what it returns). Does that make sense?

So that I can do 

```
# generate path
fields = {"foo": datetime.datetime.now(), "bar": "hello" }
path = my_template.apply_fields(fields)
# path : "/my/lovely/path/foo/2015-03-23_12_34_34"

fields = template.get_fields(path)
fields = {"foo": datetime.datetime(2015,03,23,12,34,34), "bar": "hello" }
```

Does that make sense? Can you foresee any use cases where what you are putting in is *not* what you want to get back out later on?